framework/src/modules/interoperability/utils.ts

```diff
@@ -106,3 +146,295 @@ export const computeValidatorsHash = (
+	partnerValidators: ChainValidators,
+	partnerChainAccount: ChainAccount,
+	certificate: Certificate,
+	header: BlockHeader,
+) => {
+	// Certificate and Validators Update Validity
+	const { activeValidators, certificateThreshold } = partnerValidators;
+	if (!txParams.certificate.equals(EMPTY_BYTES)) {
+		const verifySignature = verifyWeightedAggSig(
+			activeValidators.map(v => v.blsKey),
+			certificate.aggregationBits as Buffer,
+			certificate.signature as Buffer,
+			MESSAGE_TAG_CERTIFICATE.toString('hex'),
+			partnerChainAccount.networkID,
+			txParams.certificate,
+			activeValidators.map(v => v.bftWeight),
+			certificateThreshold,
+		);
+
+		if (!verifySignature || certificate.timestamp >= header.timestamp)
+			throw Error(
+				`Certificate is invalid due to invalid last certified height or timestamp or signature.`,
+			);
+	}
+};
+
+export const checkValidatorsHashWithCertificate = (
+	txParams: CrossChainUpdateTransactionParams,
+	certificate: Certificate,
+	partnerValidators: ChainValidators,
+) => {
+	if (
+		txParams.activeValidatorsUpdate.length !== 0 ||
+		txParams.newCertificateThreshold > BigInt(0)
+	) {
+		const newActiveValidators = updateActiveValidators(
+			partnerValidators.activeValidators,
+			txParams.activeValidatorsUpdate,
+		);
+		const validatorsHash = computeValidatorsHash(
+			newActiveValidators,
+			txParams.newCertificateThreshold || partnerValidators.certificateThreshold,
+		);
+
+		if (!certificate.validatorsHash.equals(validatorsHash)) {
+			throw new Error('Validators hash is incorrect given in the certificate.');
+		}
+	}
+};
+
+export const commonCCUExecutelogic = async (args: CommonExecutionLogicArgs) => {
+	const {
+		certificate,
+		partnerChainAccount,
+		partnerValidatorStore,
+		partnerChainStore,
+		partnerValidators,
+		chainIDBuffer,
+		context,
+	} = args;
+	const newActiveValidators = updateActiveValidators(
+		partnerValidators.activeValidators,
+		context.params.activeValidatorsUpdate,
+	);
+	partnerValidators.activeValidators = newActiveValidators;
+	if (context.params.newCertificateThreshold !== BigInt(0)) {
+		partnerValidators.certificateThreshold = context.params.newCertificateThreshold;
+	}
+	await partnerValidatorStore.setWithSchema(
+		chainIDBuffer,
+		partnerValidators,
+		chainValidatorsSchema,
+	);
+	if (!context.params.certificate.equals(EMPTY_BYTES)) {
+		partnerChainAccount.lastCertificate.stateRoot = certificate.stateRoot;
+		partnerChainAccount.lastCertificate.timestamp = certificate.timestamp;
+		partnerChainAccount.lastCertificate.height = certificate.height;
+		partnerChainAccount.lastCertificate.validatorsHash = certificate.validatorsHash;
+	}
+
+	await partnerChainStore.setWithSchema(chainIDBuffer, partnerChainAccount, chainAccountSchema);
+
+	const partnerChannelStore = context.getStore(
+		context.transaction.moduleID,
+		STORE_PREFIX_CHANNEL_DATA,
+	);
+	const partnerChannelData = await partnerChannelStore.getWithSchema<ChannelData>(
```

Comment:
Missing update logic for parnerChainOutboxRoot

Reply:
ah i updated it but forgot to push from my local 😅